client/blocks/qr-code-login/index.jsx

```diff
@@ -0,0 +1,185 @@
+import { getTracksAnonymousUserId } from '@automattic/calypso-analytics';
+import { Card } from '@automattic/components';
+import { useTranslate } from 'i18n-calypso';
+import QRCode from 'qrcode.react';
+import { useEffect, useState } from 'react';
+import qrCenter from 'calypso/assets/images/qr-login/wp.png';
+import { setStoredItem, getStoredItem } from 'calypso/lib/browser-storage';
+import { useInterval } from 'calypso/lib/interval';
+
+import './style.scss';
+
+const AUTH_PULL_INTERVAL = 5000; // 5 seconds
+const LOCALE_STORAGE_KEY = 'qr-login-token';
+
+const isStillValidToken = ( tokenData ) => {
+	if ( ! tokenData?.expires ) {
+		return false;
+	}
+	return tokenData.expires > Date.now() / 1000;
+};
+
+const getLoginActionResponse = async ( action, args ) => {
+	const url = new URL( 'https://wordpress.com/wp-login.php' );
+	url.searchParams.append( 'action', action );
+
+	Object.keys( args ).forEach( ( key ) => {
+		url.searchParams.append( key, args[ key ] );
+	} );
+
+	const response = await fetch( url.href );
+	return await response.json();
+};
+
+const fetchQRCodeData = async ( setTokenData, tokenData, anonymousUserId ) => {
+	if ( isStillValidToken( tokenData ) ) {
+		return tokenData;
+	}
+	// tokenData is set to null initially.
+	// Lets wait till it is set to false when the local data is the just yet.
+	if ( tokenData === null ) {
+		return null;
+	}
+
+	const responseData = await getLoginActionResponse( 'qr-code-token-request-endpoint', {
+		anon_id: anonymousUserId,
+	} );
+
+	setTokenData( responseData.data );
+	setStoredItem( LOCALE_STORAGE_KEY, responseData.data );
+};
+
+const fetchAuthState = async ( setAuthState, setTokenData, tokenData, anonymousUserId ) => {
+	if ( ! tokenData ) {
+		return;
+	}
+
+	if ( ! isStillValidToken( tokenData ) ) {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId );
+		return;
+	}
+
+	const { token, encrypted } = tokenData;
+
+	const responseData = await getLoginActionResponse( 'qr-code-authentication-endpoint', {
+		anon_id: anonymousUserId,
+		token: token,
```

Comment:
Nit: could be shortened to:

```suggestion
		token,
```